base_trgm_search/models/trgm_index.py

```diff
@@ -0,0 +1,224 @@
+# -*- coding: utf-8 -*-
+import logging
+
+from openerp import api, exceptions, fields, models
+from openerp.osv import expression
+
+from psycopg2.extensions import AsIs
+
+_logger = logging.getLogger(__name__)
+
+
+class TrgmIndex(models.Model):
+
+    """Model for Trigram Index."""
+
+    _name = 'trgm.index'
+    _rec_name = 'field_id'
+
+    field_id = fields.Many2one(
+        comodel_name='ir.model.fields',
+        string='Field',
+        required=True,
+        help='You can either select a field of type "text" or "char".'
+    )
+
+    index_name = fields.Char(
+        string='Index Name',
+        readonly=True,
+        help='The index name is automatically generated like '
+             'fieldname_indextype_idx. If the index already exists and the '
+             'index is located in the same table then this index is resused. '
+             'If the index is located in another table then a number is added '
+             'at the end of the index name.'
+    )
+
+    index_type = fields.Selection(
+        selection=[('gin', 'GIN'), ('gist', 'GiST')],
+        string='Index Type',
+        default='gin',
+        required=True,
+        help='Cite from PostgreSQL documentation: "As a rule of thumb, a '
+             'GIN index is faster to search than a GiST index, but slower to '
+             'build or update; so GIN is better suited for static data and '
+             'GiST for often-updated data."'
+    )
+
+    @api.model
+    def _trgm_extension_exists(self):
+        self.env.cr.execute("""
+            SELECT name, installed_version
+            FROM pg_available_extensions
+            WHERE name = 'pg_trgm'
+            LIMIT 1;
+            """)
+
+        extension = self.env.cr.fetchone()
+        if extension is None:
+            return 'missing'
+
+        if extension[1] is None:
+            return 'uninstalled'
+
+        return 'installed'
+
+    @api.model
+    def _is_postgres_superuser(self):
+        self.env.cr.execute("SHOW is_superuser;")
+        superuser = self.env.cr.fetchone()
+        return superuser is not None and superuser[0] == 'on' or False
+
+    @api.model
+    def _install_trgm_extension(self):
+        extension = self._trgm_extension_exists()
+        if extension == 'missing':
+            _logger.warning('To use pg_trgm you have to install the '
+                            'postgres-contrib module.')
+        elif extension == 'uninstalled':
+            if self._is_postgres_superuser():
+                self.env.cr.execute("CREATE EXTENSION IF NOT EXISTS pg_trgm;")
+                return True
+            else:
+                _logger.warning('To use pg_trgm you have to create the '
+                                'extension pg_trgm in your database or you '
+                                'have to be the superuser.')
+        else:
+            return True
```

Comment:
It would be nice to have an INFO message informing that trigram search is available.